### ISSUE TITLE
(migratus/initialize config)

### ISSUE THREAD

**Issue by MafcoCinco** [yogthos/migratus#85]
First, Migratus has been great.  Tight and simple migration tool.  

Having said that, I would like the ability to create the database itself.  Currently, AFAIK, the user must initialize the database (schema) outside the scope of the project (i.e. CREATE SCHEMA <schema_name>).  Other migration tools that I have used (specifically in the context of Ruby on Rails) provide a `setup` or `initialize` function which allows the user to generate the schema(s) from a clean DB.  Would be awesome to have in Migratus as it would provide a uniform interface for initializing the DB and make the project more self contained.

If I'm missing something and this is already possible, I apologize.

**Comment by yogthos** [comment]
I think that would be a reasonable idea. It sounds like it would just be a matter of having an `init` function that would check if the db exists, and run a SQL script if none is found. So, in addition to the standard migration files there would be something like `init.sql` for the schema creation?

**Comment by MafcoCinco** [comment]
Yep, exactly.

**Comment by yogthos** [comment]
Yeah I think that would be easy enough to add. I'll take a look when I get a chance, but if it's something you think you'd like to try to add I'm open to a pr as well.

**Comment by yogthos** [comment]
I just checked in an update that adds an `init` function for running the initialization script. It looks for the `init.sql` by default, and a custom name can be supplied using the `:init-script` key in the config. If you can try it out locally and let me know that it works as expected, I can then push out a new release with the feature.

**Comment by MafcoCinco** [comment]
I'm getting an error and I think it has to do with the connection string.  When the DB/schema is present, connection string is something like `mysql://host/schema`.  Since there isn't a schema when `init` is run, should I create a separate connection configuration for use with `init`, dropping the schema portion of the URI?

**Comment by yogthos** [comment]
Yeah, looks like you would need to use a different connection for that.

**Comment by MafcoCinco** [comment]
Yep.  Makes sense.  Need to connect without selecting a DB/Schema.  Might want to add that to the doc string and/or regular documentation.

**Comment by yogthos** [comment]
Yup, I'll update the docs once I push out the new release. So, once you use a different connection string everything is working as expected then?

**Comment by MafcoCinco** [comment]
Actually, no.  When I run the normal migration, I get an error I would expect (`Unknown database`) as the DB schema has not yet been created.  However, running with the same config minus the schema in the connection string, I get something totally different: `java.lang.IllegalArgumentException: db-spec null is missing a required parameter`.  It makes me think that the configuration object is not getting passed along correctly within Migratus for the `init` function?  

Not 100% sure but things look correct in my client code.

**Comment by yogthos** [comment]
Currently, you have to call `connect` explicitly to initialize the connection:

```clojure
(let [store (proto/make-store config)]
      (proto/connect store)
      (proto/init store)
      (proto/disconnect store))
```

Perhaps `init` should try to connect automatically though.

**Comment by MafcoCinco** [comment]
I see.  I was calling `(migratus/init config)` just like `(migratus/migrate config)` or `(migratus/rollback)`.  Looks like there is a little bit more to it currently but I agree, it should probably connect automatically like the other functions do.  Let me try your code.

**Comment by MafcoCinco** [comment]
different error but still an error :-(.  Getting `clojure.lang.PersistentVector cannot be cast to clojure.lang.Named` when calling `proto/connect`.  Is there any additional munging that needs to happen on the store object from what would normally be passed to `migratus/migrate`?

**Comment by yogthos** [comment]
Hmm, take a look [here](https://github.com/yogthos/migratus/blob/f24349abff66bb21c1f31f617c8778cbadda3b16/test/migratus/test/database.clj#L75) at how the test is setup.

**Comment by MafcoCinco** [comment]
Configuration options look correct.  Some values are different.  I'm using `mysql` instead of `h2`.  My `classname` is `com.mysql.jdbc.Driver` (using version `0.3.7` of `org.clojure/java.jdbc`).  It looks similar to [this issue](https://github.com/yogthos/migratus/issues/68) (at least the error it is reporting is the same).

**Comment by yogthos** [comment]
Yeah, clojure.java.jdbc had breaking changes in the API, and `connect`  tries to create a migration table if it doesn't exist using the new API.

**Comment by MafcoCinco** [comment]
Yeah.  Saw that in the log file in the line before the error.  Is it possible to turn off migration table creation for the `init` phase of the migration?

**Comment by yogthos** [comment]
You could just create the migration table manually. That step will be skipped when it exists.

**Comment by MafcoCinco** [comment]
I don't think I can exactly.  The migration table lives in the schema that I'm trying to create with the init file.  I could create it in the init file but that is less than ideal and, more importantly, I'm not sure it will work.  Is it testing for the existence of the migration table prior to running the init file?  Even if it works, probably don't want to force a user to create a book keeping table that is an implementation detail to the library.

**Comment by yogthos** [comment]
That's a good point actually, the `init` can't rely on the `connect` function since that one assumes a schema. I'll update to have it do its own connection.

**Comment by yogthos** [comment]
ok just pushed a fix for that

**Comment by MafcoCinco** [comment]
The schema creation was successful, so that is good.  However, when I tried to follow that with a run of my migrations, they failed with the same error the init was getting before (`PersistentVector` casting issue).

**Comment by yogthos** [comment]
Right, but that's a problem with the library version. The workaround in this case would be to create the migration table in the schema creation script.

**Comment by MafcoCinco** [comment]
Okay.  I can do that.  When you say library version, you mean the JDBC version I'm using?

**Comment by MafcoCinco** [comment]
Just tried to create `schema_migrations` table via `init.sql` with no luck.  Here is the contents of `init.sql`, which I think is correct syntax for running multiple statements in a single file:

```sql
CREATE SCHEMA database;
--;;
CREATE TABLE database.table;
```
Gives a very unhelpful (love MySQL error reporting!) SQL syntax error

**Comment by yogthos** [comment]
Yeah that part I'm not sure about. I'm not too familiar with MySQL.

**Comment by hyleung** [comment]
@MafcoCinco I was trying to do the same thing with MySQL. Oddly enough, when I put the `CREATE DATABASE` command in the `init.sql` script and use a _separate_ migration script for creating the table, it seems to work.

The multi-command syntax works fine when I do it that way, but errors when I have it in the `init.sql` script. ¯\_(ツ)_/¯

**Comment by MafcoCinco** [comment]
@hyleung Thanks for the update.  I will give that a try.

**Comment by hyleung** [comment]
@MafcoCinco btw, I should add - you might want to use a different config to create the schema vs. execute the migrations, cos there'll be a table created to track `schema_migrations` in whatever schema you connect to initially. For example, I connected using `localhost/mysql` and the tracking table was created there, which might not be desirable.

**Comment by J-ogawa** [comment]
@hyleung I met similar error at in "0.9.3" about MySQL. But no error in "0.8.13".

**Comment by dijonkitchen** [comment]
Perhaps something like templating the scripts like here: https://github.com/rails/rails/blob/master/activerecord/Rakefile#L179 ?

**Comment by lane-s** [comment]
I'm having trouble with the difference in connection strings as well. When running init.sql I need to use the subname `//localhost:5432/postgres`, but then my migrations will either be using `//localhost:5432/my_db` or `//localhost:5432/test_db`. Would it be possible to add a `:init-subname` key to the config?

**Comment by yogthos** [comment]
Yeah that sounds reasonable to me. Might be better to just have a separate optional `:init-db` key, and when it's present the connection config will be read from there rather than the one defined in `:db`. I'm not sure I'll have time to look at this in the near future, but I can help out with a PR. It should be a fairly small change.
